File: sdk/workbuddy-enterprises/python-sdk/src/workbuddy_enterprise/resources/expert_categories.py

from __future__ import annotations

from typing import Any, Sequence

from workbuddy_enterprise.resources._base import Resource
from workbuddy_enterprise.response import ApiResponse, Page


class ExpertCategoriesResource(Resource):
    def list(self) -> ApiResponse[Page[dict[str, Any]]]:
        return self._as_page(self._get("/openapi/expert-categories"))

    def create(
        self,
        *,
        name: str,
        description: str | None = None,
        sort_order: int | None = None,
    ) -> ApiResponse[dict[str, Any]]:
        body: dict[str, Any] = {"name": name}
        if description is not None:
            body["description"] = description
        if sort_order is not None:
            body["sortOrder"] = sort_order
        return self._as_map(self._post_json("/openapi/expert-categories", body=body))

    def update(
        self,
        category_id: int,
        *,
        name: str | None = None,
        description: str | None = None,
        sort_order: int | None = None,
    ) -> ApiResponse[None]:
        body: dict[str, Any] = {}
        if name is not None:
            body["name"] = name
        if description is not None:
            body["description"] = description
        if sort_order is not None:
            body["sortOrder"] = sort_order
        resp = self._post_json(f"/openapi/expert-categories/{category_id}/update", body=body)
        return ApiResponse(None, resp.code, resp.message, resp.request_id, resp.raw)

    def delete(self, category_id: int) -> ApiResponse[dict[str, Any]]:
        return self._as_map(self._post_json(f"/openapi/expert-categories/{category_id}/delete", body=None, send_json=False))

    def reorder(self, ordered_ids: Sequence[int]) -> ApiResponse[None]:
        resp = self._post_json(
            "/openapi/expert-categories/reorder",
            body={"orderedIds": list(ordered_ids)},
        )
        return ApiResponse(None, resp.code, resp.message, resp.request_id, resp.raw)

File: sdk/workbuddy-enterprises/rust-sdk/src/resources/expert_categories.rs
use crate::client::Client;
use crate::error::Result;
use crate::operations::validate_json_operation;
use crate::response::ApiResponse;
use serde_json::{json, Value};

pub struct ExpertCategoriesResource<'a> {
    pub(crate) client: &'a Client,
}

impl ExpertCategoriesResource<'_> {
    pub fn list(&self) -> Result<ApiResponse<crate::response::Page<Value>>> {
        self.client.get_page("/openapi/expert-categories", &[])
    }

    pub fn create(
        &self,
        name: &str,
        description: Option<&str>,
        sort_order: Option<i64>,
    ) -> Result<ApiResponse<Value>> {
        let mut body = serde_json::Map::new();
        body.insert("name".into(), json!(name));
        if let Some(d) = description {
            body.insert("description".into(), json!(d));
        }
        if let Some(s) = sort_order {
            body.insert("sortOrder".into(), json!(s));
        }
        let body = Value::Object(body);
        validate_json_operation("expert-categories-create", &body)?;
        self.client
            .post_json("/openapi/expert-categories", &[], body)
    }

    pub fn update(&self, category_id: i64, body: Value) -> Result<ApiResponse<Value>> {
        validate_json_operation("expert-categories-update", &body)?;
        self.client.post_json(
            &format!("/openapi/expert-categories/{category_id}/update"),
            &[],
            body,
        )
    }

    pub fn delete(&self, category_id: i64) -> Result<ApiResponse<Value>> {
        self.client.post_empty(
            &format!("/openapi/expert-categories/{category_id}/delete"),
            &[],
        )
    }

    pub fn reorder(&self, ordered_ids: &[i64]) -> Result<ApiResponse<Value>> {
        let body = json!({ "orderedIds": ordered_ids });
        validate_json_operation("expert-categories-reorder", &body)?;
        self.client
            .post_json("/openapi/expert-categories/reorder", &[], body)
    }
}
